### PR TITLE
Portal events

### PR DESCRIPTION
**Pull request by he-is-james**
- restructured the model for Events to be more specific to time and dates
- added pagination for dates as well as re-occurring events for weekly and monthly
- formats the date and time for the event cards
- can preview the events with another format
- can add new events, converting date objects to strings to store
- can edit existing events following the same idea for create events

- implemented adding, editing, deleting images
- should test creating a new event with and without images
- should test editing an existing event by removing all images, decreasing the number of images, increasing the number of images, keeping the number of images the same (upload two new and remove two existing)
- should test deleting an event with and without images

### CODE REVIEW

File: server/controllers/eventController.js [hunk 1]
```diff
@@ -17,43 +54,83 @@ const getAllEvents = async (req, res) => {
...
-      console.log(err);
-    } else {
-      res.json(data);
-    }
-  });
+  const { updatedEvent } = req.body;
+  const eventData = JSON.parse(updatedEvent);
+
+  const { files } = req;
+  const updatedImageKeys = files.map((file) => `EventImages/${file.originalname}`);
```
Comment (siddharthnandy) [inline server/controllers/eventController.js:82]: Again, unique filenames

File: server/controllers/eventController.js [hunk 2]
```diff
@@ -17,43 +54,83 @@ const getAllEvents = async (req, res) => {
...
+    await Promise.all(toDelete.map(async (key) => {
+      await s3.deleteObject({
+        Bucket: process.env.S3_BUCKET,
+        Key: key,
+      }).promise();
+    }));
+
+    // Add the image files that were added from the original
+    files.forEach(async (file) => {
+      const imageKey = `EventImages/${file.originalname}`;
```
Comment (siddharthnandy) [inline server/controllers/eventController.js:104]: unique filenames

File: server/routes/eventRoutes.js [hunk 3]
```diff
@@ -1,14 +1,18 @@
...
+const storage = multer.memoryStorage();
+const upload = multer({ storage });
+
 eventRouter.use(passport.authenticate('jwt', { session: false }));
 
-eventRouter.post('/post', eventController.createEvent);
-eventRouter.get('/get', eventController.getAllEvents);
-eventRouter.patch('/update/:id', eventController.updateEvent);
+eventRouter.post('/create', upload.array('image'), eventController.createEvent);
+eventRouter.get('/getevents', eventController.getAllEvents);
```
Comment (siddharthnandy) [inline server/routes/eventRoutes.js:14]: make sure event API change will be reflected in mobile app

File: website/src/Pages/EventsPage/EventsPage.jsx [hunk 4]
```diff
@@ -1,10 +1,249 @@
...
+              startTime={event.dateDetails.startTime}
+              endTime={event.dateDetails.endTime}
+              title={event.title}
+              host={event.host}
+              tag={event.tag}
+            />
+          </Box>
+        ))}
+      </Box>
+      <Modal
```
Comment (siddharthnandy) [inline website/src/Pages/EventsPage/EventsPage.jsx:188]: image
styling is a bit off but idk if that's a matter for this PR

**Review by siddharthnandy — CHANGES_REQUESTED** [review]

**Review by siddharthnandy — APPROVED** [review]

File: website/src/Pages/EventsPage/AddEventPage.jsx [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+  const [address, setAddress] = useState('');
+  const [description, setDescription] = useState('');
+  const [imageData, setImageData] = useState([]);
+  const [imagePreviews, setImagePreviews] = useState([]);
+
+  const backToEvents = () => {
+    navigate('/events');
+  };
+
+  const createEvent = async () => {
```
Comment (siddharthnandy) [inline website/src/Pages/EventsPage/AddEventPage.jsx:37]: validation might be a good idea for some of these fields, with popups for bad fields (i.e., non-numeric lat/long)

**Review by siddharthnandy — CHANGES_REQUESTED** [review]

File: website/src/Pages/EventsPage/AddEventPage.jsx (hunk 5 above)
Comment (siddharthnandy) [reply website/src/Pages/EventsPage/AddEventPage.jsx:37]: (made this after starting Resource PR)

Comment (siddharthnandy) [reply website/src/Pages/EventsPage/AddEventPage.jsx:37]: also EditEvents page too

Comment (he-is-james) [reply website/src/Pages/EventsPage/AddEventPage.jsx:37]: I think we can keep this for the styling edit? @zhangjuliet let me know if you want me to implement this

**Review by siddharthnandy — APPROVED** [review]

**Pull request merged**